### ISSUE TITLE
Survey recent face identification model

### ISSUE THREAD

**Issue by kyakuno** [axinc-ai/ailia-models#1392]
現状、arcfaceとinsightfaceしかないので、商用に耐えうる最新のモデルがないかなと考えています。

**Comment by mucunwuxian** [comment]
取り急ぎ、slackにて共有させて頂いた内容
https://axincai.slack.com/archives/C019HCVQBCP/p1710974187099469?thread_ts=1707350225.508149&cid=C019HCVQBCP

**Comment by mucunwuxian** [comment]
**📝 （Githubリポジトリ確認Rough）**

# Fast and Interpretable Face Identification for Out-Of-Distribution Data Using Vision Transformers

- official Implementation : https://github.com/anguyen8/face-vit

<br/>

## 記載のデータやモデル等について

- 学習とテストを行うためのデータセットの紹介として、幾つかのリンクを用意して下さっている
    - https://github.com/anguyen8/face-vit?tab=readme-ov-file#2-download-datasets-and-pretrained-models
    - この内、CASIAデータセットは紹介のリンクからDownloadすることが難しいが、別途Kaggleのリンクからデータを得ることができる
        - https://www.kaggle.com/datasets/debarghamitraroy/casia-webface
    - LFWについては、out-of-distribution（OOD）とするための加工がされた「マスク」「サングラス」のデータなどをダウンロードすることができる
        - test.pyを動かす際には、比較対象ペアを表現したtxtファイルが必要だが、それは見つけることができず、自前で作成をして精度評価を実施した
            - データを用意して、 `itertools.combinations(filename_list, 2)` 等を駆使して、比較的簡単にリストは作成できる
    - 別途、以下リンクのAsian Face Dataset Cleanedが有望そうなのだが、zipファイルのパスワードが分からず、使用できなかった
        - https://github.com/vitoralbiero/afd_dataset_cleaned?tab=readme-ov-file
        - READMEに沿って、パスワード共有のお願いメールをしたが、返信なし
    - 別途以下のリンクから、Asian-celeb datasetというものをダウンロードできる
        - https://github.com/deepinsight/insightface/issues/1977
        - 100GB弱程の容量の、個人毎にフォルダ分けされた学習データ「train_celebrity.tar.gz」をダウンロードすることができた
        - 解凍すると、以下のようなレイアウトであった
            - ./celebrity
                ├── 10000127
                │   ├── 1174212268,1428908528_align.jpg
                │   ├── 1471363678,3088703426_align.jpg
                │   ├── 1567547479,2287369560_align.jpg
                │   ├── …
                │   ├── b8014a90f603738d505b34b7b31bb051f819ec16.jpg_align.jpg
                │   ├── bd3eb13533fa828b053aa576fd1f4134970a5a0d.jpg_align.jpg
                │   └── d043ad4bd11373f0632e570aa40f4bfbfbed0415.jpg_align.jpg
                ├── 10000207
                │   ├── 1517662845,2845745492_align.jpg
                │   ├── 2120650568,2478378549_align.jpg
                │   ├── 2396100771,3332616371_align.jpg
                │   ├── 4014354810,993095686_align.jpg
                │   ├── 4279193289,1758596758_align.jpg
                │   ├── d52a2834349b033be23d51f611ce36d3d539bd61.jpg_align.jpg
                │   └── fab3ac11717bde3eca80c48e.jpg_align.jpg
                ├── 10000614
                │   ├── 1068056581,531401204_align.jpg
                │   ├── 1293933361,2638801324_align.jpg
                │   ├── 3174911707,2119134179_align.jpg
                │   ├── 3659451043,3944349514_align.jpg
                │   ├── 4119301830,799137584_align.jpg
                │   └── 65190419,3888778856_align.jpg
                ├── 10000787
                │   ├── 2009960157,2531342572_align.jpg
                │   ├── 3128351165,680612430_align.jpg
                │   ├── 3141727179,462710453_align.jpg
                │   ├── 4146439009,708557137_align.jpg
                │   └── 912703374,3883341723_align.jpg
                ├── …
                └── 9999882
                    ├── 1166746776,1838900629_align.jpg
                    ├── 1277534803,3776517958_align.jpg
                    ├── 1341285813,3213518941_align.jpg
                    ├── 2559246026,1872684710_align.jpg
                    ├── 2594604176,2083631403_align.jpg
                    ├── 2642547371,2863194642_align.jpg
                    ├── 3397068385,1319670352_align.jpg
                    ├── 373907790,285742577_align.jpg
                    ├── 3847435221,2656644531_align.jpg
                    ├── 4bd1e8039361b86b3812bb2e.jpg_align.jpg
                    ├── 5af4d7ead81d409fd439c905.jpg_align.jpg
                    ├── 7aad4ae7c0035962b8382004.jpg_align.jpg
                    ├── caae6809a19ddae32fddd45d.jpg_align.jpg
                    ├── d8b8c92aeefa8813d52af147.jpg_align.jpg
                    ├── ea85a9453005f46acefca35e.jpg_align.jpg
                    └── faacb56469e30ebcf636545d.jpg_align.jpg
                93979 directories, 2830146 files
        - 以下のような、解像度400x400の画像が含まれる
            - ![2745349,580661141_align](https://github.com/axinc-ai/ailia-models/assets/9916906/454c586b-3291-409d-b018-2ca78791b2db)
            - ![20313288,3215213190_align](https://github.com/axinc-ai/ailia-models/assets/9916906/388c4048-f972-40e9-9c8c-30d390d83f20)
            - ![4173634,2961012115_align](https://github.com/axinc-ai/ailia-models/assets/9916906/c0ebd3d5-f0eb-4933-a4d3-1c2322d21cbc)

<br/>

- リポジトリのリンクより、幾つか用意してくれている学重済みモデルファイルに辿り着くことができるが、各ファイル内容についての説明がリポジトリ上になく、かつ、論文に記載されている内容ともリンクさせづらくはある
    - test.pyで使用できるモデルは、論文中に記載されているH2L（Hybrid-ViT with 2-image input）という型だと思われる
        - ViT-P8S8_2-image_webface_2m_arcface_resnet18_s1_depth_4_head_8_hybrid/best.pt
        - ViT-P8S8_2-image_webface_2m_arcface_resnet18_s1_depth_8_head_2_hybrid/best.pt
        - etc...
    - H2Lモデルは、論文ではcosine similarityを用いる旨が記載されているが、test.pyではEuclid-Distanceが用いられている
      <img width="257" alt="image" src="https://github.com/axinc-ai/ailia-models/assets/9916906/12e18949-0e50-4ecc-bd84-6027e1921c45">
        - また、train.pyは、lossの計算については、cross-entropyのみのオプションとなっている
        - Euclid-Distanceとcross-entropyとで学習をした結果だとは、どうも考えづらそうである為、学習コードはコーディングの必要がありそうな印象
        - embedding特徴を算出する迄の処理はtest.pyに記載がされている為、[pytorch-metric-learning](https://github.com/KevinMusgrave/pytorch-metric-learning/tree/master/examples)などを参考に実装すれば、学習の実施は可能

<br/>

## 日本人の顔認証でのテスト精度

- 顔が似ているとされる芸能人のペアを幾つかダウンロードする形で、テストデータを各人20画像ずつ準備
    - ダウンロードをした方
        - 三上愛さんx小松菜奈さん
        - 広瀬アリスさんx水野美紀さん
        - 宮崎あおいさんx二階堂ふみさんx山本舞香さん
        - 工藤大輝さんx山下健二郎さん
        - 田中圭さんx向井理さん
        - 竹野内豊さんx大谷亮平さん
        - 福士蒼汰さんx中川大志さん
    - 同傾向の画像を避けるようにしてダウンロードを実施
    - ダウンロードした画像に対して、dlibによる顔検出を実施し、得られたBoundingBoxを長辺で正方形に加工したうえで、10%のpaddingを付与し、そこからcrop画像を得た
        - Asian-celeb datasetと比べると、位置合わせや、未補正の傾きが、若干キツイ印象
- モデルは幾つか適用を試みたが、以下のモデルにて、Euclid距離（L2）を適用した際が最も精度が高かった
    - ViT-P8S8_2-image_webface_2m_arcface_resnet18_s1_depth_4_head_8_hybrid/best.pt
        - LFWで学習されたモデルであり、基本は欧米人の認証に強い形
        - LFWのデータからラフに精度検証を行ってみたところで、accuracyが0.95を超えることが多かった
        - 論文記載によれば、99%を超えるとのこと
    - Cosine距離も用意されていたので試してみたが、精度低く、恐らくはモデルの学習がそれによって行われていない様子であった
    - 精度は、最もaccuracyが高い閾値にて、以下confusion-matrixであった
        - tp:0.06 fp0.07
           fn:0.02 tn0.85
           acc:0.9089086859688196
        - 予測分布（青が比較の2画像間の距離、橙が同一人物が1で異なる人物が0のラベル値）
           ![image](https://github.com/axinc-ai/ailia-models/assets/9916906/d8f78449-9ebd-410c-b781-68c9fb1ddb5d)
            - 苦戦している様子が伺える
        - 欧米人の顔画像にて学習した結果にしては、精度が高いのかもしれない
        - 似ているとされる芸能人のペアを多く含めた為、少し厳しい評価になっている可能性がある
        - 識別誤りの傾向分析までは、まだできていない
            - 顔の傾きや、オクルージョンなども影響している可能性があると考えている
        - 男性は男性、女性は女性とだけ比較をしている
            - 同性については、似ているとされるペア以外にも組み合わせ（計22,450ペア）にて、認証比較を実施した

<br/>
